### PR TITLE
ENH: Added loggamma entropy method

### PR DESCRIPTION
**Pull request by MatteoRaso**
#### Reference issue

Part of #17832 
#### What does this implement/fix?

Adds a formula for the loggamma entropy, as well as an asymptotic expansion for larger values.
#### Additional information

I wasn't able to get mpmath to generate test values with integration, so if anybody else could get it to work, that'd be helpful. There was discussion in previous PRs about using sympy or Wolfram Alpha to help get the asymptotic expansion for entropies. Since the expansion for this method is already very accurate, I decided that wasn't necessary. Attached to this PR is a writeup explaining how I derived the entropy formula and the asymptotic expansion.
loggamma
[loggamma_entropy.pdf](https://github.com/scipy/scipy/files/10984729/loggamma_entropy.pdf)

### CODE REVIEW

**Comment by MatteoRaso** [comment]
@mdhaber The changes you requested in #18122 are in this PR.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:1678]:
```suggestion
        assert_allclose(stats.loggamma.entropy(c), ref, rtol=1e-14)
```

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:1693]: As the generic tests agree with the analytical formula you derived (all those new results you derived could be noted somewhere else than scipy also ), I think we can provide test reference values using an mpmath implementation of that formula. This is not really a test of the asymptotic formula at the moment.

Please also move all tests in just one function. If reference values are calculated in the same way for all tested values, there is no point in different testing functions in my opinion.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:5847]:
```suggestion
        h = _lazywhere(c >= 45, (c, ), f=asymptotic, f2=regular)
```

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Nice result (very accurate expansion early on), just the tests need another look.

File: scipy/stats/tests/test_distributions.py
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py:1693]: >all those new results you derived could be noted somewhere else than scipy btw like on arXiv

ArXiv probably isn't the best place to put them, since it's not really something you would ever publish in a journal and I don't think you can add to the article once you submit it. That being said, I do plan to compile them at some point for easier reading, kind of like that entropy book you posted a few months ago.

>If reference values are calculated in the same way for all tested values, there is no point in different testing functions in my opinion.

They weren't calculated the same way. For the first test, I just called the entropy method before changing anything, saved it, and then copied it into the test function. Also, the tolerances are different for the 2 tests, which is probably enough justification to keep them separate.

**Comment by mdhaber** [comment]
@mdhaber The changes you requested in https://github.com/scipy/scipy/pull/18122 are in this PR.

Thank you! But

> There was discussion in previous PRs about using sympy or Wolfram Alpha to help get the asymptotic expansion for entropies. Since the expansion for this method is already very accurate, I decided that wasn't necessary.

I thought I should rephrase my previous comments about this (https://github.com/scipy/scipy/pull/18135#issuecomment-1468869071, https://github.com/scipy/scipy/pull/18094#issuecomment-1467440742): the purpose of using sympy or Wolfram Alpha is not to get _more accurate_ asymptotic expansions. Rather, it is to simplify development, avoid (reasonably accurate but nonetheless) incorrect expansions, and speed up review. You don't need to follow this suggestion, but I wouldn't want it to be dismissed due to a misunderstanding.

For example, instead of undertaking the derivation by hand and preparing the PDF, it may have saved you time to begin with the `sympy` code you referred to (https://github.com/scipy/scipy/pull/18094#issuecomment-1467440742). By adding the symbol `c` instead of `df` and redefinining:

```python3
old =  gammaln(c) - c * digamma(c) + c
```
the script pops out an answer:

```python3
-log(c)/2 + log(2)/2 + 1/2 + log(pi)/2 + 1/(6*c) - 1/(90*c**3)
```

<details>
(If you want the `c**-5` term, just add it to the expansion of `gammaln` before executing the code. There is a balance between the accuracy and complexity of adding extra terms, and where to stop is a matter of choice. I got the impression before that you did not want to add terms past $1/c^4$ (https://github.com/scipy/scipy/pull/18131#issuecomment-1469002212), and that's fine. Either way, I don't think it undermines the point of the PR.)
</details>

Also, by letting the computer do the symbol manipulation, we avoid the possibility of mistakes in the algebra. Sometimes these mistakes are subtle - we've seen this a few times (https://github.com/scipy/scipy/pull/18094#issuecomment-1467440742, https://github.com/scipy/scipy/pull/18122#issuecomment-1467473461, https://github.com/scipy/scipy/pull/17874#discussion_r1089880639). For instance, the expansion can give _good_ results even when there are mistakes, but results are even better when the math is correct. Moreover, committing implementations of mathematically incorrect formulas can erode confidence in the rest of SciPy, so I prefer to avoid it when possible.

Finally, although I believed that the derivation was correct in this case (because it was better than the default implementation for `x` as small as ~40), I still need to check it before it can be merged. For me, it's easier to review the line
```
old =  gammaln(c) - c * digamma(c) + c
```
than it is to review a two-page PDF, despite its careful preparation.

There may be other reviewers with different preferences, but this is my suggestion to help you get your contributions reviewed efficiently.

**Comment by MatteoRaso** [comment]
>Rather, it is to simplify development, avoid (reasonably accurate but nonetheless) incorrect expansions, and speed up review.

Okay, now I understand. I'll consider using computer assistance to validate my expansions, but I'm still going to do them by hand as well, just in case.

> I got the impression before that you did not want to add terms past 1 / c^4

That was because the PR in question was for an issue that specifically focused on large values. If you focus too much on adding more terms, the PR stops becoming about large values and more about getting the most accurate asymptotic expansion. I know that's pretty pedantic, but I think it's good to make sure every PR sticks to a well-defined goal.

**Comment by mdhaber** [comment]
> but I think it's good to make sure every PR sticks to a well-defined goal.

Absolutely. I'm sorry for the misunderstanding, but in that discussion, I didn't ask for there to be more terms. I hoped to clear that up when I [wrote](https://github.com/scipy/scipy/pull/18131#issuecomment-1469088478):

> No, I didn't suggest adding more terms.

**Comment by MatteoRaso** [comment]
>I'm sorry for the misunderstanding

No big deal, these things happen. At least we're now on the same page.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:1693]:
```suggestion
    @pytest.mark.parametrize("c, ref",
                             [(1e-4, 10.210440363752314),
                              (1, 1.5772156649015328),
                              (1e4, -3.186214986116763),
                              (1e10, -10.093986931748889),
                              (1e100, -113.71031611649761)])
    def test_entropy(self, c, ref):
    
        # Reference values were calculated with mpmath
        # from mpmath import mp
        # mp.dps = 500
        # def loggamma_entropy_mpmath(c):
        #       c = mp.mpf(c)
        #       return float(mp.log(mp.gamma(c)) + c * (mp.one - mp.digamma(c)))
        
        assert_allclose(stats.loggamma.entropy(c), ref, rtol=1e-14)
```

**Review by dschmitz89 — CHANGES_REQUESTED** [review]

**Comment by dschmitz89** [comment]
This PR fixes the entropy for small and large `c`:

image

<details>

```
from mpmath import mp
from scipy import special as sc
import numpy as np
import matplotlib.pyplot as plt

mp.dps = 500 

@np.vectorize
def loggamma_entropy_mpmath(c):
    c = mp.mpf(c)
    return float(mp.log(mp.gamma(c)) + c * (mp.one - mp.digamma(c)))

def loggamma_entropy(c):
    return np.where(c < 45,
                    sc.gammaln(c) - c * sc.digamma(c) + c,
                    -0.5 * np.log(c) + (1 / 6) * c ** -1. -(1 / 90) * c ** -3. + (1 / 210) * c ** -5. + 0.5 + 0.5 * np.log(2 * np.pi))

c = np.logspace(-8, 20, 1000)
plt.semilogx(c, stats.loggamma.entropy(c), label = "main", ls='dotted')
plt.semilogx(c, loggamma_entropy(c), label = "PR", ls='dashdot')
plt.semilogx(c, np.array(loggamma_entropy_mpmath(c), np.float64), label = "mpmath", ls='dashed')
plt.legend()
plt.show()
```
</details>

I propose to commit my suggestions to simplify the tests and then merge it as it is a definite improvement.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py:1693]: All cases but `1` fail in the main branch.

**Comment by MatteoRaso** [comment]
Ok, I've merged the suggestion made by @dschmitz89.

**Review by dschmitz89 — APPROVED** [review]
The linter is complaining about lines not touched in this PR, so better not waste another round of CI minutes on that.

**Comment by MatteoRaso** [comment]
@mdhaber any updates on this?

**Comment by mdhaber** [comment]
I didn't plan to review this further right now. Someone else is welcome to.

**Comment by MatteoRaso** [comment]
>I didn't plan to review this further right now. Someone else is welcome to.

@dschmitz89 reviewed this 2 weeks ago and approved the changes, along with the recommendation to merge this PR once I committed his suggestions.

**Comment by tylerjereddy** [comment]
I discussed this a bit with Pamphile and Matt in person and I think the decision was to bump the milestone at this point, or at least I didn't get a clear indication that we were comfortable merging at this point.

I'll bump the milestone and happily merge once folks are convinced we're ready to put this in.

**Comment by steppi** [comment]
It looks like I missed the ping and this is still waiting for a review from me. I’ll try to get to it today.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5837]: Mathematica agrees.

image

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5844]: The usual script agrees.
<details>

```python3
import sympy
from sympy import log, pi, Rational, simplify, expand
from IPython.display import display
x, c = sympy.symbols('x, c')

def digamma(x):
    return log(x) - 1/(2*x) - 1/(12*x**2) + 1/(120*x**4) - 1/(252*x**6)

def gammaln(x):
    return x*log(x) - x - log(x)/2 + log(2*pi)/2 + 1/(12*x) - 1/(360*x**3) + 1/(1260*x**5)

old = gammaln(c) - c * digamma(c) + c

display(expand(simplify(old)))
```

</details>

image

**Comment by mdhaber** [comment]
Re: https://github.com/scipy/scipy/pull/18156#issuecomment-1474758635, I don't think we need to worry about negative `c` because `loggamma` is not defined for negative `c`.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5844]: Might as well use the same style as above and simplify.
```suggestion
            term = -0.5*np.log(c) + c**-1./6 - c**-3./90 + c**-5./210
            h = norm._entropy() + term
```

**Comment by mdhaber** [comment]
Failures so far are unrelated (CircleCi, Linux Meson Python Debug & ATLAS)

**Review by mdhaber — APPROVED** [review]

**Pull request merged**
